cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1233 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"blainsmith.com/go/seahash"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/cdcpb"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/kv/regionlock"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/tablepb"
+	"github.com/pingcap/tiflow/pkg/chann"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/pdutil"
+	"github.com/pingcap/tiflow/pkg/spanz"
+	"github.com/pingcap/tiflow/pkg/txnutil"
+	"github.com/pingcap/tiflow/pkg/util"
+	"github.com/pingcap/tiflow/pkg/version"
+	"github.com/prometheus/client_golang/prometheus"
+	kvclientv2 "github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/tikv"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	grpcerror "google.golang.org/grpc/codes"
+	grpcmeta "google.golang.org/grpc/metadata"
+)
+
+// SubscriptionID comes from `SharedClient.AllocSubscriptionID`.
+type SubscriptionID uint64
+
+// MultiplexingEvent is like model.RegionFeedEvent.
+type MultiplexingEvent struct {
+	model.RegionFeedEvent
+	SubscriptionID SubscriptionID
+	Start          time.Time
+}
+
+// SharedClient is shared in many tables. Methods are thread-safe.
+type SharedClient struct {
+	changefeed model.ChangeFeedID
+	config     *config.KVClientConfig
+	metrics    sharedClientMetrics
+
+	clusterID  uint64
+	filterLoop bool
+
+	pd           pd.Client
+	grpcPool     GrpcPool
+	regionCache  *tikv.RegionCache
+	pdClock      pdutil.Clock
+	lockResolver txnutil.LockResolver
+
+	// requestRangeCh is used to retrieve subscribed table tasks.
+	requestRangeCh *chann.DrainableChann[rangeTask]
+	// regionCh is used to cache region tasks after ranges locked.
+	regionCh *chann.DrainableChann[singleRegionInfo]
+	// regionRouter is used to cache region tasks with rpcCtx attached.
+	regionRouter *chann.DrainableChann[singleRegionInfo]
+	// resolveLockCh is used to retrieve resolve lock tasks.
+	resolveLockCh *chann.DrainableChann[resolveLockTask]
+	errCh         *chann.DrainableChann[regionErrorInfo]
+
+	workers []*sharedRegionWorker
+
+	// only modified in requestRegionToStore so lock is unnecessary.
+	requestedStores map[string]*requestedStore
+
+	totalSpans struct {
+		sync.RWMutex
+		m *spanz.HashMap[SubscriptionID]
+		v map[SubscriptionID]*requestedTable
```

Comment:
Fields in the struct are confused, so it is preferred to use meaningful names.

Reply:
`m` is removed now.